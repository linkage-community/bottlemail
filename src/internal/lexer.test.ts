import tokenize, { TokenType } from "./lexer"

describe("Lexer", () => {
	const expectedATToken: TokenType = {
		type: "AT",
		payload: "@"
	}
	const expectedCOLONToken: TokenType = {
		type: "COLON",
		payload: ":"
	}

	describe("Type 'AT' cases", () => {
		const test1times = "@"
		test(`'${test1times}' must have 1 length. All tokens must be 'AT' type.`, () => {
			const tokens = tokenize(test1times)
			expect(tokens).toEqual([expectedATToken])
		})

		const test5times = test1times.repeat(5)
		test(`'${test5times}' must have 5 length. All tokens must be  'AT' type.`, () => {
			const tokens = tokenize(test5times)
			expect(tokens).toEqual(new Array(5).fill(expectedATToken))
		})
	})

	describe("Type 'COLON' cases", () => {
		const test1times = ":"
		test(`'${test1times}' must have 1 length. All tokens must be 'COLON' type.`, () => {
			const tokens = tokenize(test1times)
			expect(tokens).toEqual([expectedCOLONToken])
		})

		const test5times = test1times.repeat(5)
		test(`'${test5times}' must have 5 length. All tokens must be 'COLON' type.`, () => {
			const tokens = tokenize(test5times)
			expect(tokens).toEqual(new Array(5).fill(expectedCOLONToken))
		})
	})

	describe("Type 'WHITESPACE' cases", () => {
		const space = " "
		test(`SPACE must have 1 length. All tokens must be 'TEXT' type.`, () => {
			const tokens = tokenize(space)
			const expected: TokenType[] = [
				{
					type: "WHITESPACE",
					payload: space
				}
			]
			expect(tokens).toEqual(expected)
		})

		const tab = "\t"
		test(`TAB must have 1 length. All tokens must be 'TEXT' type.`, () => {
			const tokens = tokenize(tab)
			const expected: TokenType[] = [
				{
					type: "WHITESPACE",
					payload: tab
				}
			]
			expect(tokens).toHaveLength(1)
			expect(tokens).toEqual(expected)
		})

		const lf = "\n"
		test(`LF must have 1 length. All tokens must be 'WHITESPACE' type.`, () => {
			const tokens = tokenize(lf)
			const expected: TokenType = {
				type: "WHITESPACE",
				payload: lf
			}
			expect(tokens).toHaveLength(1)
			expect(tokens[0]).toEqual(expected)
		})

		const long = [space.repeat(5), lf.repeat(3), tab.repeat(3)].join("")
		test(`long whitespaces (${long}) must have 1 length. All tokens must be 'WHITESPACE' type.`, () => {
			const tokens = tokenize(long)
			const expected: TokenType[] = [
				{
					type: "WHITESPACE",
					payload: long
				}
			]
			expect(tokens).toEqual(expected)
		})
	})

	describe("Type 'TEXT' cases", () => {
		const test1 = "This=is=text!!!~~HAPPY~~YEAH!"
		test(`'${test1}' must have 1 length. All tokens must be 'TEXT' type.`, () => {
			const tokens = tokenize(test1)
			const expected: TokenType[] = [
				{
					type: "TEXT",
					payload: test1
				}
			]
			expect(tokens).toEqual(expected)
		})
	})

	describe("Complex cases", () => {
		const testcase1 =
			"@otofune Yo! :smile: https://github.com/ http://[fe80::a1b3:125d:c1f8:4780]/ @ @test"
		test(`'${testcase1}' must equal to snapshot`, () => {
			const tokens = tokenize(testcase1)
			expect(tokens).toMatchSnapshot()
		})

		const testcase2 = "@ :smile:"
		test(`'${testcase2}' must be expected`, () => {
			const tokens = tokenize(testcase2)
			const expected: TokenType[] = [
				expectedATToken,
				{
					type: "WHITESPACE",
					payload: " "
				},
				expectedCOLONToken,
				{
					type: "TEXT",
					payload: "smile"
				},
				expectedCOLONToken
			]
			expect(tokens).toEqual(expected)
		})
	})
})
