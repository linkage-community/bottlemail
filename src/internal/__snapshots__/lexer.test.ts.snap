// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lexer Complex cases '@otofune Yo! :smile: https://github.com/ http://[fe80::a1b3:125d:c1f8:4780]/ @ @test' must equal to snapshot 1`] = `
Array [
  Object {
    "payload": "@",
    "type": "AT",
  },
  Object {
    "payload": "otofune",
    "type": "TEXT",
  },
  Object {
    "payload": " ",
    "type": "WHITESPACE",
  },
  Object {
    "payload": "Yo!",
    "type": "TEXT",
  },
  Object {
    "payload": " ",
    "type": "WHITESPACE",
  },
  Object {
    "payload": ":",
    "type": "COLON",
  },
  Object {
    "payload": "smile",
    "type": "TEXT",
  },
  Object {
    "payload": ":",
    "type": "COLON",
  },
  Object {
    "payload": " ",
    "type": "WHITESPACE",
  },
  Object {
    "payload": "https",
    "type": "TEXT",
  },
  Object {
    "payload": ":",
    "type": "COLON",
  },
  Object {
    "payload": "//github.com/",
    "type": "TEXT",
  },
  Object {
    "payload": " ",
    "type": "WHITESPACE",
  },
  Object {
    "payload": "http",
    "type": "TEXT",
  },
  Object {
    "payload": ":",
    "type": "COLON",
  },
  Object {
    "payload": "//[fe80",
    "type": "TEXT",
  },
  Object {
    "payload": ":",
    "type": "COLON",
  },
  Object {
    "payload": ":",
    "type": "COLON",
  },
  Object {
    "payload": "a1b3",
    "type": "TEXT",
  },
  Object {
    "payload": ":",
    "type": "COLON",
  },
  Object {
    "payload": "125d",
    "type": "TEXT",
  },
  Object {
    "payload": ":",
    "type": "COLON",
  },
  Object {
    "payload": "c1f8",
    "type": "TEXT",
  },
  Object {
    "payload": ":",
    "type": "COLON",
  },
  Object {
    "payload": "4780]/",
    "type": "TEXT",
  },
  Object {
    "payload": " ",
    "type": "WHITESPACE",
  },
  Object {
    "payload": "@",
    "type": "AT",
  },
  Object {
    "payload": " ",
    "type": "WHITESPACE",
  },
  Object {
    "payload": "@",
    "type": "AT",
  },
  Object {
    "payload": "test",
    "type": "TEXT",
  },
]
`;
